### ISSUE TITLE
Syntax for data-providers

### ISSUE THREAD

**Issue by redaxmedia** [phpbench/phpbench#188]
The syntax of the data-provider is confusing while dealing with multiple parameter - you should stick to PHPUnit and provide $param1, $param2 and $param3 instead of the single $params array.

Here comes the example from the PHPUnit documentation:

<pre>
<?php
class DataTest extends PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider additionProvider
     */
    public function testAdd($a, $b, $expected)
    {
        $this->assertEquals($expected, $a + $b);
    }

    public function additionProvider()
    {
        return array(
          array(0, 0, 0),
          array(0, 1, 1),
          array(1, 0, 1),
          array(1, 1, 3)
        );
    }
}
?>
</pre>

**Comment by dantleech** [comment]
So in PHPBench data providers look like

````
public function fooProvider()
{
    return array(
        array(
            'nb_properties' => 6,
            'algorithm' => 'foo',
        )
    );
}
````

i.e. they are (optionally) associative. this is mainly because it allows us to combine multiple providers and iterate over the cartesian product (see [documentation](http://phpbench.readthedocs.org/en/latest/writing-benchmarks.html#parameterized-benchmarks)).

What we could, perhaps, do is match the array keys of the data provided to argument names on the function:

````php
function benchFoo($bar_foo, $foo_bar)
{
}

function provideFoo()
{
    return array(array('bar_foo' => 1, 'foo_bar' => 2));
}
````

And throw an exception if the data is an assocative array or a standard array (detecting the difference is fun).

But still, I kind of prefer having a limited arity on the benchmarking functions - it leaves room for being flexible in the future, rather than dictating that all arguments must be used for parameters always.

**Comment by redaxmedia** [comment]
All right, let me give you a real world example...

1. The data provider for my Hash class
https://github.com/redaxmedia/redaxscript/blob/3.0.0/tests/provider/hash.json

2. The PHPUnit test
https://github.com/redaxmedia/redaxscript/blob/3.0.0/tests/includes/HashTest.php#L135-L161

3. The PHPBench bench
https://github.com/redaxmedia/redaxscript/blob/3.0.0/benchs/includes/HashBench.php#L67-L86


I hope you can see that your approach is not easy to understand - even on such a simple bench

**Comment by dantleech** [comment]
Well, I cannot argue my side here definitively, both aproaches have the advantages and disadvantages.

The problems with expanding the array keys to arguments as I see it are:

- Adding arguments to the benchmark subject affects call time (last time I checked) - maybe that doesn't matter as it would be a constant.
- It makes order difficut to figure out when using multiple parameter providers.
- It makes editing the data provider data more difficult and prone to error (it also means that each set of data is coupled to each other set - requiring the same number and same order of array elements) - I have encountered this issue in PHPUnit.

Additional pros for using a single array argument:

- Its simple.

And cons:

- If you don't use an associative array, you are faced with having to remember the numerical indexes of your data.

Of those, the second is probably the bigest disadvantage, and it is one that might be solved with the method i indicated above.
